lib/classes/ConfigSchemaHandler/errorMessageBuilder.js

```diff
@@ -0,0 +1,169 @@
+    return ajvErrors;
+  }
+
+  const anyOfErrorIndex = ajvErrors.findIndex(
+    ajvError => ajvError.keyword === 'anyOf' && ajvError.schemaPath === ANY_OF_EVENT_SCHEMA_PATH
+  );
+
+  // Alternatively, if there is a deeper error, then it's better to make it parent
+  // because it actually the reason why other errors are generated by AJV
+  const deeperAnyOfErrorIndex = ajvErrors.findIndex(ajvError =>
+    DEEPER_ANY_OF_SUB_ERROR_REGEX.test(ajvError.schemaPath)
+  );
+
+  const finalIndex = deeperAnyOfErrorIndex >= 0 ? deeperAnyOfErrorIndex : anyOfErrorIndex;
+
+  const result = [ajvErrors[finalIndex]];
+  result[0].params.suberrors = [];
+  ajvErrors.forEach((ajvError, index) => {
+    const isAnyOfSubError = ANY_OF_SUB_ERROR_REGEX.test(ajvError.schemaPath);
+
+    if (deeperAnyOfErrorIndex >= 0) {
+      if (isAnyOfSubError || index === anyOfErrorIndex) {
+        result[0].params.suberrors.push(ajvError);
+      } else if (index !== deeperAnyOfErrorIndex) {
+        result.push(ajvError);
+      }
+    }
+
+    if (deeperAnyOfErrorIndex < 0) {
+      if (isAnyOfSubError) {
+        result[0].params.suberrors.push(ajvError);
+      } else if (index !== anyOfErrorIndex) {
+        result.push(ajvError);
+      }
+    }
+  });
+
+  return result;
+};
+
+/*
+ * For error object structure, see https://github.com/ajv-validator/ajv/#error-objects
+ */
+const buildErrorMessages = (ajvErrors, userConfig = {}) => {
+  let errors = ajvErrors;
+
+  errors = structureAjvErrors(errors);
+
+  errors = addUserFriendlyMessage(errors, userConfig);
+
+  errors = errors.sort((a, b) => b.dataPath.length - a.dataPath.length);
+  if (!errors.length) throw new Error('Validation errors array cannot be emptry');
+
+  return errors.map(error => {
+    if (error.friendlyMessage) {
+      return error.friendlyMessage;
+    }
+    return buildDefaultErrorMessage(error);
+  });
+};
+
+const buildDefaultErrorMessage = ajvError => {
+  switch (ajvError.keyword) {
+    case 'additionalProperties': {
+      let additionalProperty;
+      if (ajvError.params && ajvError.params.additionalProperty) {
+        additionalProperty = ajvError.params.additionalProperty;
+      }
+      const formattedDataPath = getFormattedDataPath(ajvError.dataPath);
+      return `Unrecognized property '${additionalProperty}' on '${formattedDataPath}'`;
+    }
+    default:
+      return `${getFormattedDataPath(ajvError.dataPath)} ${ajvError.message}`;
+  }
+};
+
+const addUserFriendlyMessage = (errors, userConfig) => {
+  return errors.map(error => {
+    let friendlyMessage;
+
+    // regex matches '.functions['xxx'].events[x].xxx'
+    const isEventParam = /^\.functions\['[\w-_]+'\]\.events\[[0-9]+\]\.[\w]+$/.test(error.dataPath);
+    if (error.params && error.params.additionalProperty && isEventParam) {
+      const eventName = error.dataPath.split('.')[3];
+      friendlyMessage = `Unsupported parameter '${error.params.additionalProperty}' for '${eventName}' event`;
+    }
```

Comment:
I would remove that generalization. I think it's fine if error appears as:

```
Unrecognized property 'x' on 'functions.x.events[0].x'
```

It'll be even better as it provides clear information for which function error happened and at which event (by index).